ocular/src/keyring.rs

```diff
@@ -0,0 +1,740 @@
+        })
+    }
+
+    fn get_all_keys(&self) -> Result<Vec<PublicKeyOutput>, Box<dyn std::error::Error>> {
+        let mut vec = Vec::new();
+
+        for entry in fs::read_dir(&self.key_path).expect("Could not read directory.") {
+            let path = entry.unwrap().path();
+
+            if path.is_file() {
+                if let Some(extension) = path.extension() {
+                    if extension == "pem" {
+                        let name = path
+                            .file_stem()
+                            .expect("Could not get file stem.")
+                            .to_str()
+                            .expect("Could not convert to string.");
+                        let key_data = self.get_key_address(name)?;
+
+                        vec.push(key_data);
+                    }
+                }
+            }
+        }
+
+        Ok(vec)
+    }
+
+    fn recover_from_mnemonic(
+        &self,
+        name: &str,
+        mnemonic: &str,
+        password: &str,
+        derivation_path: Option<&str>,
+        override_if_exists: bool,
+    ) -> Result<(), KeyStoreError> {
+        // Check if key already exists
+        if self.key_exists(name)? && !override_if_exists {
+            eprintln!("Key '{}', already exists.", name);
+            return Err(KeyStoreError::Exists(name.to_string()));
+        }
+
+        let seed = Seed::new(
+            &bip39::Mnemonic::from_phrase(mnemonic.trim(), bip39::Language::English)
+                .expect("Could not read mnemonic."),
+            password,
+        );
+
+        let derivation_path = match derivation_path {
+            Some(_i) => derivation_path.unwrap(),
+            _ => COSMOS_BASE_DERIVATION_PATH,
+        };
+
+        let derivation_path = derivation_path
+            .parse::<bip32::DerivationPath>()
+            .expect("Could not parse derivation path.");
+
+        // Process key and store
+        let key =
+            bip32::XPrv::derive_from_path(seed, &derivation_path).expect("Could not derive key.");
+        let private_key = k256::SecretKey::from(key.private_key());
+        let encoded_key = private_key
+            .to_pkcs8_der()
+            .expect("Could not PKCS8 encode private key");
+
+        let key_name = &KeyName::new(name)
+            .unwrap_or_else(|_| panic!("Could not create KeyName for '{}'.", name));
+        self.key_store
+            .as_ref()
+            .expect("Error accessing key store.")
+            .store(key_name, &encoded_key)
+            .expect("Could not store key");
+
+        Ok(())
+    }
+}
+
+// ---------------------------------- Tests ----------------------------------
+// TODO: Make these tests more comprehensive and increase code coverage.
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn file_key_store_without_path_init_test() {
+        let keyring = Keyring::new_file_store(Option::None);
+
+        assert_eq!(keyring.key_store.key_store_created(), true);
```

Comment:
The way I'm viewing it is, "will a failure in this test tell me exactly what unit failed or just that something failed in general?", the former being a unit test, the latter being integration. If the pieces (denoted by each assertion) give enough info to make it obvious what line of what method failed then I think it's ok to keep it as a unit test, otherwise IMO it should be integration.

Reply:
That's fair. My goal here was to split these tests up method wise (which would be the former), however it does indeed rely on quite a few setup steps & other methods. Happy to move these to the integration test area, however I'm wondering if there's actually any way to unit test some of these code paths in addition to that.